File: native/android/core-model/src/main/kotlin/com/fireball/nativeapp/core/model/ArtistReleaseProbe.kt
package com.fireball.nativeapp.core.model

/**
 * Pure helpers for followed-artist iTunes release polling (Android + iOS parity).
 */
object ArtistReleaseProbe {
    data class AlbumHit(
        val collectionId: String,
        val collectionName: String,
    )

    enum class ReleaseOutcome {
        /** No library row change. */
        Unchanged,
        /** First baseline stored (no user notification). */
        BaselineStored,
        /** New release detected; notify channels. */
        NotifyNewRelease,
    }

    data class ReleaseEvaluation(
        val artist: Artist,
        val outcome: ReleaseOutcome,
    )

    fun resolveNumericItunesArtistId(
        storedArtistId: String,
        artistName: String,
        lookupByName: (String) -> Int?,
    ): Int? =
        storedArtistId.toIntOrNull() ?: lookupByName(artistName.trim())

    fun evaluateNewRelease(
        artist: Artist,
        resolvedArtistId: String,
        latestAlbum: AlbumHit?,
    ): ReleaseEvaluation {
        var working = artist
        if (resolvedArtistId != artist.artistId) {
            working = artist.copy(artistId = resolvedArtistId)
        }
        val hit = latestAlbum ?: return ReleaseEvaluation(working, ReleaseOutcome.Unchanged)
        if (hit.collectionId == working.latestReleaseId) {
            return ReleaseEvaluation(working, ReleaseOutcome.Unchanged)
        }
        val updated = working.copy(latestReleaseId = hit.collectionId)
        return if (working.latestReleaseId == null) {
            ReleaseEvaluation(updated, ReleaseOutcome.BaselineStored)
        } else {
            ReleaseEvaluation(updated, ReleaseOutcome.NotifyNewRelease)
        }
    }

    fun notificationCopy(artistName: String, albumName: String): Pair<String, String> {
        val title = "New Release from $artistName"
        val message = "$albumName is out now!"
        return title to message
    }
}

File: native/ios/FireballNative/Core/ArtistReleaseProbe.swift
import Foundation

/// Pure helpers for followed-artist iTunes release polling (Android + iOS parity).
enum ArtistReleaseProbe {
    struct AlbumHit: Equatable {
        let collectionId: String
        let collectionName: String
    }

    enum ReleaseOutcome: Equatable {
        case unchanged
        case baselineStored
        case notifyNewRelease
    }

    struct ReleaseEvaluation: Equatable {
        let artist: Artist
        let outcome: ReleaseOutcome
    }

    static func resolveNumericItunesArtistId(
        storedArtistId: String,
        artistName: String,
        lookupByName: (String) -> Int?
    ) -> Int? {
        if let stored = Int(storedArtistId.trimmingCharacters(in: .whitespacesAndNewlines)) {
            return stored
        }
        return lookupByName(artistName.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    static func evaluateNewRelease(
        artist: Artist,
        resolvedArtistId: String,
        latestAlbum: AlbumHit?
    ) -> ReleaseEvaluation {
        var working = artist
        if resolvedArtistId != artist.artistId {
            working = Artist(
                artistId: resolvedArtistId,
                name: artist.name,
                artwork: artist.artwork,
                latestReleaseId: artist.latestReleaseId
            )
        }
        guard let hit = latestAlbum else {
            return ReleaseEvaluation(artist: working, outcome: .unchanged)
        }
        if hit.collectionId == working.latestReleaseId {
            return ReleaseEvaluation(artist: working, outcome: .unchanged)
        }
        let updated = Artist(
            artistId: working.artistId,
            name: working.name,
            artwork: working.artwork,
            latestReleaseId: hit.collectionId
        )
        if working.latestReleaseId == nil {
            return ReleaseEvaluation(artist: updated, outcome: .baselineStored)
        }
        return ReleaseEvaluation(artist: updated, outcome: .notifyNewRelease)
    }

    static func notificationCopy(artistName: String, albumName: String) -> (title: String, message: String) {
        ("New Release from \(artistName)", "\(albumName) is out now!")
    }
}
